examples/read-write-transactions.rs

```diff
@@ -0,0 +1,112 @@
+use firestore::{errors::FirestoreError, paths, FirestoreDb};
+use futures::stream::FuturesOrdered;
+use serde::{Deserialize, Serialize};
+use tokio_stream::StreamExt;
+
+pub fn config_env_var(name: &str) -> Result<String, String> {
+    std::env::var(name).map_err(|e| format!("{}: {}", name, e))
+}
+
+// Example structure to play with
+#[derive(Debug, Clone, Deserialize, Serialize)]
+struct MyTestStructure {
+    counter: i64,
+}
+
+const TEST_COLLECTION_NAME: &'static str = "test";
+const TEST_DOCUMENT_ID: &str = "test_doc_id";
+
+/// Creates a document with a counter set to 0 and then concurrently executes futures for `COUNT_ITERATIONS` iterations.
+/// Finally, it reads the document again and verifies that the counter matches the expected number of iterations.
+#[tokio::main]
+async fn main() -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
+    // Logging with debug enabled
+    let subscriber = tracing_subscriber::fmt()
+        .with_env_filter("firestore=debug")
+        .finish();
+    tracing::subscriber::set_global_default(subscriber)?;
+
+    // Create an instance
+    let db = FirestoreDb::new(&config_env_var("PROJECT_ID")?).await?;
+
+    const COUNT_ITERATIONS: i64 = 50;
+
+    println!("Creating initial document...");
+
+    // Remove if it already exists
+    db.fluent()
+        .delete()
+        .from(TEST_COLLECTION_NAME)
+        .document_id(TEST_DOCUMENT_ID)
+        .execute()
+        .await?;
+
+    // Let's insert some data
+    let my_struct = MyTestStructure { counter: 0 };
+
+    db.fluent()
+        .insert()
+        .into(TEST_COLLECTION_NAME)
+        .document_id(TEST_DOCUMENT_ID)
+        .object(&my_struct)
+        .execute()
+        .await?;
+
+    println!("Running transactions...");
+
+    let mut futures = FuturesOrdered::new();
+
+    for _ in 0..COUNT_ITERATIONS {
+        futures.push_back(increment_counter(&db));
+    }
+
+    let results = futures.collect::<Vec<_>>().await;
+    dbg!(results);
+
+    println!("Testing results...");
+
+    let test_structure: MyTestStructure = db
+        .fluent()
+        .select()
+        .by_id_in(TEST_COLLECTION_NAME)
+        .obj()
+        .one(TEST_DOCUMENT_ID)
+        .await?
+        .expect("Missing document");
+
+    assert_eq!(test_structure.counter, COUNT_ITERATIONS);
+
+    Ok(())
+}
+
+async fn increment_counter(db: &FirestoreDb) -> Result<(), FirestoreError> {
```

Comment:
Btw, Firestore supports actually atomic document transformations  in transactions:

```
#[derive(Clone, PartialEq, ::prost::Oneof)]
        pub enum TransformType {
            /// Sets the field to the given server value.
            #[prost(enumeration="ServerValue", tag="2")]
            SetToServerValue(i32),
            /// Adds the given value to the field's current value.
            ///
            /// This must be an integer or a double value.
            /// If the field is not an integer or double, or if the field does not yet
            /// exist, the transformation will set the field to the given value.
            /// If either of the given value or the current field value are doubles,
            /// both values will be interpreted as doubles. Double arithmetic and
            /// representation of double values follow IEEE 754 semantics.
            /// If there is positive/negative integer overflow, the field is resolved
            /// to the largest magnitude positive/negative integer.
            #[prost(message, tag="3")]
            Increment(super::super::Value),
            /// Sets the field to the maximum of its current value and the given value.
            ///
            /// This must be an integer or a double value.
            /// If the field is not an integer or double, or if the field does not yet
            /// exist, the transformation will set the field to the given value.
            /// If a maximum operation is applied where the field and the input value
            /// are of mixed types (that is - one is an integer and one is a double)
            /// the field takes on the type of the larger operand. If the operands are
            /// equivalent (e.g. 3 and 3.0), the field does not change.
            /// 0, 0.0, and -0.0 are all zero. The maximum of a zero stored value and
            /// zero input value is always the stored value.
            /// The maximum of any numeric value x and NaN is NaN.
            #[prost(message, tag="4")]
            Maximum(super::super::Value),
            /// Sets the field to the minimum of its current value and the given value.
            ///
            /// This must be an integer or a double value.
            /// If the field is not an integer or double, or if the field does not yet
            /// exist, the transformation will set the field to the input value.
            /// If a minimum operation is applied where the field and the input value
            /// are of mixed types (that is - one is an integer and one is a double)
            /// the field takes on the type of the smaller operand. If the operands are
            /// equivalent (e.g. 3 and 3.0), the field does not change.
            /// 0, 0.0, and -0.0 are all zero. The minimum of a zero stored value and
            /// zero input value is always the stored value.
            /// The minimum of any numeric value x and NaN is NaN.
            #[prost(message, tag="5")]
            Minimum(super::super::Value),
            
```

They are not fully supported through high level API in this library, but still supported using a bit low-leveled:
`transaction.add()`

Reply:
Interesting, I wasn't aware of this! I suppose this makes my provided example a bit contrived. :D Perhaps I'll keep the example for now, but mention that users should prefer using `Increment` for atomic counters.

Regarding increment, here's some [documentation](https://firebase.google.com/docs/firestore/solutions/counters) and a [usage example](https://cloud.google.com/firestore/docs/samples/firestore-data-set-numeric-increment) for anybody else stumbling upon this thread.

Reply:
Just do something more fancy - divide it or multiple :D

Reply:
Or sum two different fields